pytensor/tensor/rewriting/elemwise.py

```diff
@@ -380,6 +380,99 @@ def is_dimshuffle_useless(new_order, input):
     return is_useless
 
 
+@node_rewriter([Elemwise])
+def local_elemwise_lift_scalars(fgraph, node):
+    op = node.op
+
+    if not isinstance(op, Elemwise):
+        return False
+
+    if not all(input.ndim == 0 for input in node.inputs):
+        return False
+
+    scalars = [aes.as_scalar(input) for input in node.inputs]
+
+    # TODO Something like
+    # copy_stack_trace(node.outputs[0], new_res)
+    return [as_tensor_variable(out) for out in op.scalar_op.make_node(*scalars).outputs]
+
+
+compile.optdb["specialize"].register(
+    "local_elemwise_lift_scalars",
+    local_elemwise_lift_scalars,
+    "fast_run_numba",
+    "fast_compile_numba",
+)
+
+
+@node_rewriter([Elemwise])
+def push_elemwise_constants(fgraph, node):
+    """Push constant scalars from inputs to elemwise to inputs of the
+    contained scalar op.
+    """
+    op = node.op
+
+    if not isinstance(op, Elemwise):
+        return False
+
+    if any(op.inplace_pattern):
```

Comment:
It's fine to be inplace because constants are never inplaced. But to not have to deal with it just register this rewrite before the inplace rewrites

Reply:
That shouldn't happen. Inplace rewrites are myopic, they only look at 1 node at a time. I never saw a rewrite checking inplace patterns elsewhere.

Reply:
Hm, what about examples like this?
```python
import pytensor.tensor as pt
import pytensor

x = pt.dvector("x")
y = 2 * x + 1

val = np.ones(3)
input = pytensor.In(x, update=y)
func = pytensor.function([input], [])
pytensor.dprint(func)
```
```
Elemwise{Composite{(1.0 + (2.0 * i0))}}[(0, 0)] [id A] 0
 |x [id B]
```
Replacing the inplace Elemwise with a non-inplace Elemwise would be incorrect here.
Still not a problem because the rewrite is registered before the inplace pass, but still...

Reply:
Why is the update supposed to create a problem? I mean if you are worried about this rewrite ignoring inplacing you would have to be worried every other rewrite we have the library. What is special about your push constants rewrite?

Reply:
I'm not 100% sure there would be...
But it looks like we need this final Elemwise in the graph not for its output but only for its side effect of changing the first input. If we were to replace this node with an Elemwise without the inplace flag, but the same output, wouldn't the update break? But maybe there is a feature somewhere that prevents this?